### PR TITLE
Add ReceiverChannel

### PR DESCRIPTION
**Pull request by AdrianSoundy**
## Description
Adds RMT receiver channel

## How Has This Been Tested?

Using new HC-SR04 sample (PR)

## Types of changes

- [x] Improvement (non-breaking change that improves a feature, code or algorithm)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Config and build (change in the configuration and build system, has no impact on code or features)
- [ ] Dependencies (update dependencies and changes associated, has no impact on code or features)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [CONTRIBUTING](https://github.com/nanoframework/.github/blob/master/CONTRIBUTING.md) document.
- [x] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: nanoFramework.Hardware.Esp32.Rmt/ReceiverChannel.cs [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+			get => _receiveTimeout;
+			set => _receiveTimeout = value;
+		}
+		/// <summary>
+		/// Public constructor to create receiver channel object.
+		/// </summary>
+		/// <param name="Gpio">The GPIO pin number that we want to use for receiving</param>
+		/// <param name="RmtBufferSize">The maximum number of RMT commands to be reserved in receiver input buffer. 
+		/// Default 100.</param>
+		public ReceiverChannel(int Gpio, int RmtBufferSize = 100)
```
Comment (josesimoes) [inline nanoFramework.Hardware.Esp32.Rmt/ReceiverChannel.cs:38]: parameter names: start with lower case
(same on other methods below)

File: nanoFramework.Hardware.Esp32.Rmt/ReceiverChannel.cs [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+
+		/// <summary>
+		/// Set the Idle Threshold in ticks.
+		/// </summary>
+		/// <Remarks>
+		/// The receive process finishes(goes idle) when no edges have been detected for Idle Threshold clock cycles.</Remarks>
+		/// <param name="Threshold">Value between 1 and 65535 (0xFFFF)</param>
+		public void SetIdleThresold(UInt16 Threshold)
+		{
+			if (Threshold == 0)
```
Comment (josesimoes) [inline nanoFramework.Hardware.Esp32.Rmt/ReceiverChannel.cs:83]: open and close curly braces always

File: nanoFramework.Hardware.Esp32.Rmt/RmtCommand.cs [hunk 3]
```diff
@@ -3,111 +3,114 @@
...
+		/// <summary>
+		/// Command level 1
+		/// </summary>
+		private bool _level0 = true;
 
-        /// <summary>
-        /// Command level 2
-        /// </summary>
-        private bool _level1 = false;
+		/// <summary>
```
Comment (josesimoes) [inline nanoFramework.Hardware.Esp32.Rmt/RmtCommand.cs:16]: indentation with spaces, please

**Review by josesimoes — CHANGES_REQUESTED** [review]
LGTM, just please fix the code style related stuff

File: nanoFramework.Hardware.Esp32.Rmt/TransmitterChannel.cs [hunk 4]
```diff
@@ -390,64 +331,58 @@ public void Dispose()
         #region native calls
 
         [MethodImpl(MethodImplOptions.InternalCall)]
-        private static extern void NativeInit(int gpio);
```
Comment (josesimoes) [inline nanoFramework.Hardware.Esp32.Rmt/TransmitterChannel.cs:393]: All these native calls should be `static extern`... Is there a reason for not being `static`?

File: nanoFramework.Hardware.Esp32.Rmt/TransmitterChannel.cs [hunk 5]
```diff
@@ -390,64 +331,58 @@ public void Dispose()
...
+        private extern void NativeDispose();
 
         //[MethodImpl(MethodImplOptions.InternalCall)]
-        //private static extern int NativeGetChannelStatus(int channel);
-
-        [MethodImpl(MethodImplOptions.InternalCall)]
-        private static extern byte NativeGetClockDivider();
+        //private extern int NativeGetChannelStatus(int channel);
 
         //[MethodImpl(MethodImplOptions.InternalCall)]
```
Comment (josesimoes) [inline nanoFramework.Hardware.Esp32.Rmt/TransmitterChannel.cs:372]: Better to remove all these commented methods. Let's keep the code as clean and tidy as possible.

File: nanoFramework.Hardware.Esp32.Rmt/TransmitterChannel.cs (hunk 4 above)
Comment (AdrianSoundy) [reply nanoFramework.Hardware.Esp32.Rmt/TransmitterChannel.cs:393]: @josesimoes 
When static i was having exceptions with accessing the fields in the class. Especially when trying to access the base class RmtChannel fields. It was like the pThis pointer was wrong.
I have seen this before.  I probably what you woud expect for a static method in a class.

File: nanoFramework.Hardware.Esp32.Rmt/TransmitterChannel.cs (hunk 5 above)
Comment (AdrianSoundy) [reply nanoFramework.Hardware.Esp32.Rmt/TransmitterChannel.cs:372]: These were all existing commented out calls. Didn't really look much on the transmit side to see if they were required or not. I can remove if you think they should be.

Comment (josesimoes) [reply nanoFramework.Hardware.Esp32.Rmt/TransmitterChannel.cs:372]: Yes please. Let's keep the code as clean and tidy as possible.

File: nanoFramework.Hardware.Esp32.Rmt/TransmitterChannel.cs (hunk 4 above)
Comment (josesimoes) [reply nanoFramework.Hardware.Esp32.Rmt/TransmitterChannel.cs:393]: Curious... If it's working without `static` let it be then.

File: nanoFramework.Hardware.Esp32.Rmt/TransmitterChannel.cs (hunk 5 above)
Comment (AdrianSoundy) [reply nanoFramework.Hardware.Esp32.Rmt/TransmitterChannel.cs:372]: @josesimoes , I have reviewed the transmitter code and i can see that some of the commented out code would be useful but would rather just leave for the moment as some of it needs to move to base class. Will come back to it when i have time to do some more samples that show its use.
So if you are happy can we merge as is ?

**Comment by josesimoes** [comment]
Sure. Go ahead!

**Pull request merged**